crates/sui-framework/packages/sui-framework/sources/crypto/bls12381.move

```diff
@@ -18,4 +22,227 @@ module sui::bls12381 {
+
+    public fun g2_from_bytes(bytes: &vector<u8>): Element<G2> {
+        group_ops::from_bytes(G2_TYPE, bytes, false)
+    }
+
+    public fun g2_identity(): Element<G2> {
+        let identity = G2_IDENTITY_BYTES;
+        group_ops::from_bytes(G2_TYPE, &identity, true)
+    }
+
+    public fun g2_generator(): Element<G2> {
+        let generator = G2_GENERATOR_BYTES;
+        group_ops::from_bytes(G2_TYPE, &generator, true)
+    }
+
+    public fun g2_add(e1: &Element<G2>, e2: &Element<G2>): Element<G2> {
+        group_ops::add(G2_TYPE, e1, e2)
+    }
+
+    public fun g2_sub(e1: &Element<G2>, e2: &Element<G2>): Element<G2> {
+        group_ops::sub(G2_TYPE, e1, e2)
+    }
+
+    public fun g2_mul(e1: &Element<Scalar>, e2: &Element<G2>): Element<G2> {
+        group_ops::mul(G2_TYPE, e1, e2)
+    }
+
+    /// Returns e2 / e1, fails if scalar is zero.
+    public fun g2_div(e1: &Element<Scalar>, e2: &Element<G2>): Element<G2> {
+        group_ops::div(G2_TYPE, e1, e2)
+    }
+
+    public fun g2_neg(e: &Element<G2>): Element<G2> {
+        g2_sub(&g2_identity(), e)
+    }
+
+    /// Hash using DST = BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_NUL_
+    public fun hash_to_g2(m: &vector<u8>): Element<G2> {
+        group_ops::hash_to(G2_TYPE, m)
+    }
+
+    /// Let 'scalars' be the vector [s1, s2, ..., sn] and 'elements' be the vector [e1, e2, ..., en].
+    /// Returns s1*e1 + s2*e2 + ... + sn*en.
+    /// Aborts with `EInputTooLong` if the vectors are larger than 32 (may increase in the future).
+    public fun g2_multi_scalar_multiplication(scalars: &vector<Element<Scalar>>, elements: &vector<Element<G2>>): Element<G2> {
+        group_ops::multi_scalar_multiplication(G2_TYPE, scalars, elements)
+    }
+
+    /////////////////////////////////
+    ////// Gt group operations //////
+    
+    public fun gt_identity(): Element<GT> {
+        let identity = GT_IDENTITY_BYTES;
+        group_ops::from_bytes(GT_TYPE, &identity, true)
+    }
+
+    public fun gt_generator(): Element<GT> {
+        let generator = GT_GENERATOR_BYTES;
+        group_ops::from_bytes(GT_TYPE, &generator, true)
+    }
+
+    public fun gt_add(e1: &Element<GT>, e2: &Element<GT>): Element<GT> {
+        group_ops::add(GT_TYPE, e1, e2)
+    }
+
+    public fun gt_sub(e1: &Element<GT>, e2: &Element<GT>): Element<GT> {
+        group_ops::sub(GT_TYPE, e1, e2)
+    }
+
+    public fun gt_mul(e1: &Element<Scalar>, e2: &Element<GT>): Element<GT> {
+        group_ops::mul(GT_TYPE, e1, e2)
+    }
+
+    /// Returns e2 / e1, fails if scalar is zero.
+    public fun gt_div(e1: &Element<Scalar>, e2: &Element<GT>): Element<GT> {
+        group_ops::div(GT_TYPE, e1, e2)
+    }
+
+    public fun gt_neg(e: &Element<GT>): Element<GT> {
+        gt_sub(&gt_identity(), e)
+    }
+
+    /////////////////////
+    ////// Pairing //////
+
+    public fun pairing(e1: &Element<G1>, e2: &Element<G2>): Element<GT> {
+        group_ops::pairing(G1_TYPE, e1, e2)
```

Comment:
what does it mean to be G1_TYPE pairing?

Reply:
that's an internal identifier used in the interfaces between Move and Rust

Reply:
Ah I see, originally expecting it to be `GT1`, but if it's for the curve, and there is only one pairing type, then it makes sense.